furi/core/timer.c

```diff
@@ -67,17 +67,29 @@ void furi_timer_free(FuriTimer* instance) {
 
     callb = (TimerCallback_t*)pvTimerGetTimerID(hTimer);
 
-    furi_check(xTimerDelete(hTimer, portMAX_DELAY) == pdPASS);
-
-    while(furi_timer_is_running(instance)) furi_delay_tick(2);
-
+    /* If callback memory was dynamically allocated, there is potential
+     * for a race condition between issuing xTimerDelete, the timer actually
+     * being freed, and freeing the allocated TimerID that could lead to a
+     * use-after-free situation. Therefore, we first need to set the TimerID
+     * to NULL; this is wrapped in a critical section in vTimerSetTimerID, and
+     * a NULL poiner is safely handled in TimerCallback. Then, free the
+     * allocated callback memory, and finally delete the timer.
+     * If the timer were to expire at any point along this chain and be serviced
+     * there is no risk of use-after-free of either the callback memory or the
+     * timer handle itself.
+     */
     if((uint32_t)callb & 1U) {
+        /* Set TimerID to NULL to not have a use-after-free */
+        vTimerSetTimerID(hTimer, NULL);
+
         /* Callback memory was allocated from dynamic pool, clear flag */
         callb = (TimerCallback_t*)((uint32_t)callb & ~1U);
 
         /* Return allocated memory to dynamic pool */
         free(callb);
```

Comment:
vTimerSetTimerID doesn't guarantee that timer thread is not running this callback right now. If it is then there will be condition race where you freed callb memory before it was processed.

Reply:
So you still have to ensure that timer stopped before removing callback

Reply:
That is a great point. Looks like the correct way to handle teardown in that case be to stop the timer, wait for it to be no longer active, free the callback memory, and then issue a delete request.

I'll get that updated, do some testing, and push a change to the PR in the next day or so.